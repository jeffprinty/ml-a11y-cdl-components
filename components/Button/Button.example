<div>
  <MLButton title='Primary ' primary />
  <MLButton title='Secondary' />
  <MLButton title='Styled' style={{ backgroundColor: Colors.medium_orange }} />
  <hr/>
  <MLButton icon='edit' title='Edit' primary />
  <MLButton icon='arrow_left' title='Arrow' primary />
  <MLButton icon='cancel' title='Cancel' />
</div>