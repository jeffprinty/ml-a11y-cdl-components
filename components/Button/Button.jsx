import React, { Component } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { MLIcon } from '../index';
import Colors from '../colors';
import SafeAnchor from '../SafeAnchor';

const ButtonStyle = styled.div`
  position: relative;
  display: inline-block;

  font-family: 'Source Sans Pro',sans-serif;
  font-size: 15px;
  line-height: 33px;
  color: ${props => props.primary ? Colors.pure_white : Colors.aqua};
  
  height: 34px;
  padding: ${props => props.icon ? '0 15px 0 12px' : '0 15px'};
  margin: 5px;
  
  background-color: ${props => props.primary ? Colors.aqua : 'transparent'}

  border: ${props => props.primary ? 'none' : `1px solid ${Colors.aqua}`};
  border-radius: 3px;
  
  text-decoration: none;
  cursor: pointer;
  
  &:hover {
    opacity: 0.8;
  }

  &:disabled {
    cursor: default
  }

  &:focus {
    outline: none;
    box-shadow: 0 0 6px rgba(0,117,142,1);
  }
`;
const Title = styled.span`
  line-height: 30px;
  margin-left: ${props => props.icon ? '8px' : '0'};
`;
const Icon = styled.span`
  position: relative;
  top: 4px;
  vertical-align: middle;
  color: ${props => props.primary ? Colors.pure_white : Colors.aqua};
`;

class Button extends Component {
  renderAnchor() {
    return (
      <SafeAnchor { ...this.props } />
    );
  }
  render() {
    const { onClick, icon, title, primary, href } = this.props;

    if (href) {
      return this.renderAnchor(...this.props);
    }

    return (
      <ButtonStyle
        { ...this.props }
        onClick={ onClick }
      >
        { icon ?
          <Icon>
            <MLIcon
              title={ title }
              type={ icon }
              fill={ primary ? Colors.pure_white : Colors.aqua }
            />
          </Icon>
        : null }
        <Title { ...this.props }>{ title }</Title>
      </ButtonStyle>
    );
  }
}

Button.defaultProps = {
  btnClass: 'default',
  btnType: '',
  href: '',
  icon: '',
  onClick: () => {},
  primary: false
};

Button.propTypes = {
  title: PropTypes.string.isRequired,
  onClick: PropTypes.func,
  primary: PropTypes.bool,
  href: PropTypes.string,
  icon: PropTypes.string
};

export default Button;
