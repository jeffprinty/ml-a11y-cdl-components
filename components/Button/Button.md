### Usage

Buttons provide the user with a clear way to perform an action. Buttons should be labelled clearly and in Title Case (Ex. "Sign Up")

### Special Buttons

Special buttons are alternate styles of the primary and secondary buttons used in special cases. Special buttons use color to draw attention to a specific action the button may perform. Any accessible color combination in the palette can be used as a special button.

### Icon Buttons

Icon buttons contain an additional icon to help represent the action of the button. Icons should reinforce the button action and never cause confusion. 