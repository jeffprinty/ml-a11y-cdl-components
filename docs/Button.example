<div>
  <h3>Primary Button</h3>
  <Button title="Button" onClick={ () => {} } primary />
  <Button title="Button" onClick={ () => {} } primary disabled />
  <h3>Standard Button</h3>
  <Button title="Button" onClick={ () => {} } />
  <Button title="Button" onClick={ () => {} } disabled />
</div>
