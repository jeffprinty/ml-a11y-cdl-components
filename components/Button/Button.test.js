/* eslint-disable */
import React from 'react';
import { mount, shallow, render } from 'enzyme';
import Button from './Button';

describe('/components/Button', () => {
  it('should have props for title and onClick', () => {
    const wrapper = shallow(<Button title="Button" />);
    expect(wrapper.props().title).toBe.defined;
    expect(wrapper.props().onClick).toBe.defined;
  });
  it('should render an a tag if button is not disabled', () => {
    const wrapper = mount(<Button title="Button" />);
    console.log("wrapper", wrapper);
  });
  it('should trigger onClick function when clicked', () => {
    const clickMock = jest.fn();
    const wrapper = mount(<Button title="alert" onClick={ clickMock } />);
    wrapper.simulate('click');
    expect(clickMock).toHaveBeenCalled();
  });
});
