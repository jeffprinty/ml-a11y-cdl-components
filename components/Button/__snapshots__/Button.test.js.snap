// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/components/Button should update style on hover 1`] = `
<Button
  ariaLabelledBy="testButton"
  btnClass="default"
  disabled={false}
  href={null}
  icon={null}
  onClick={[Function]}
  primary={null}
  title="Button"
>
  <styled.a
    href={null}
    icon={null}
    onClick={[Function]}
    onKeyDown={[Function]}
    primary={null}
    role="button"
    tabIndex={0}
  >
    <a
      className="bvcQgl"
      href={null}
      icon={null}
      onClick={[Function]}
      onKeyDown={[Function]}
      role="button"
      tabIndex={0}
    >
      <styled.span
        icon={null}
        id="testButton"
      >
        <span
          className="kdFLFV"
          icon={null}
          id="testButton"
        >
          Button
        </span>
      </styled.span>
    </a>
  </styled.a>
</Button>
`;

exports[`/components/Button should update style on hover 2`] = `
<Button
  ariaLabelledBy="testButton"
  btnClass="default"
  disabled={false}
  href={null}
  icon={null}
  onClick={[Function]}
  primary={null}
  title="Button"
>
  <styled.a
    href={null}
    icon={null}
    onClick={[Function]}
    onKeyDown={[Function]}
    primary={null}
    role="button"
    tabIndex={0}
  >
    <a
      className="bvcQgl"
      href={null}
      icon={null}
      onClick={[Function]}
      onKeyDown={[Function]}
      role="button"
      tabIndex={0}
    >
      <styled.span
        icon={null}
        id="testButton"
      >
        <span
          className="kdFLFV"
          icon={null}
          id="testButton"
        >
          Button
        </span>
      </styled.span>
    </a>
  </styled.a>
</Button>
`;

exports[`/components/Button with href should match the snapshot when disabled 1`] = `
<Button
  ariaLabelledBy="testButton"
  btnClass="default"
  disabled={true}
  href="http://fastgood.cheap/"
  icon={null}
  onClick={[Function]}
  primary={null}
  title="Button"
>
  <styled.button
    aria-disabled={true}
    aria-labelledby="testButton"
    ariaLabelledBy="testButton"
    btnClass="default"
    disabled={true}
    href="http://fastgood.cheap/"
    icon={null}
    onClick={[Function]}
    primary={null}
    tabIndex={0}
    title="Button"
  >
    <button
      aria-disabled={true}
      aria-labelledby="testButton"
      className="fuYbFQ"
      disabled={true}
      href="http://fastgood.cheap/"
      icon={null}
      onClick={[Function]}
      tabIndex={0}
      title="Button"
    >
      <styled.span
        icon={null}
        id="testButton"
      >
        <span
          className="kdFLFV"
          icon={null}
          id="testButton"
        >
          Button
        </span>
      </styled.span>
    </button>
  </styled.button>
</Button>
`;

exports[`/components/Button with href should match the snapshot when enabled 1`] = `
<Button
  ariaLabelledBy="testButton"
  btnClass="default"
  disabled={false}
  href="http://fastgood.cheap/"
  icon={null}
  onClick={[Function]}
  primary={null}
  title="Button"
>
  <styled.a
    href="http://fastgood.cheap/"
    icon={null}
    onClick={[Function]}
    onKeyDown={[Function]}
    primary={null}
    tabIndex={0}
  >
    <a
      className="bvcQgl"
      href="http://fastgood.cheap/"
      icon={null}
      onClick={[Function]}
      onKeyDown={[Function]}
      tabIndex={0}
    >
      <styled.span
        icon={null}
        id="testButton"
      >
        <span
          className="kdFLFV"
          icon={null}
          id="testButton"
        >
          Button
        </span>
      </styled.span>
    </a>
  </styled.a>
</Button>
`;
